### PR TITLE
Migrate deployment list page and details page actions to use new extensions

### PR DESCRIPTION
**Pull request by rohitkrai03**
**Fixes**: 
- https://issues.redhat.com/browse/ODC-6102
- https://issues.redhat.com/browse/ODC-4412

**Analysis / Root cause**:
- Deployment details page and list page used the hard coded action items.
- The HPA actions were not supported very well on details page and list page.
- Action providers had no way of adding a new action context with new scope.
- `ActionLoader` loads all actions eagerly which might impact performance on a list page where hundreds of resources would try to load actions.

**Solution Description**: 
- Migrated all the factory actions and common actions to new extension format.
- Created action providers for deployment that contribute all the actions related to deployment.
- Fixed the issue of HPA actions on details page and list page.
- Updated `ActionsLoader` to also work with reosurce action extensions.
- Contributed various action providers from different plugins using dynamic extensions.
- Updated list page and details page of deployments to use new extension actions instead of older hard coded actions.
- Created a new `ActionServiceProvider` that supports multiple contexts.
- Added support for action adapters that could update context and scope in action loader.
- Created a lazy loaded action menu that loads all the actions on hover of the action toggle.

**Screen shots / Gifs for design review**: 

**Unit test coverage report**: 

**Test setup:**

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [x] Edge

### CODE REVIEW

File: frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx:55]: This seems like an odd practice. Wouldn't this run the risk of throwing an error and crashing the page? We do it in other places which is why I said it was an odd practice.

Also, without doing anything with the response, I'm a bit confused on how this is doing anything 🤔

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx
Comment (andrewballantyne) [reply frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx:55]: Upon further thought --- I guess this is async, so it wouldn't blow up the main thread if it fails... still unsure how it helps though.

**Review by andrewballantyne — COMMENTED** [review]
There are duplicate actions now, right? Those in the wheelhouse of extensions and the original ones. Did we want to remove the old ones along with this change?

File: frontend/packages/console-app/src/actions/creators/common-factory.ts
Comment (christianvogt) [inline frontend/packages/console-app/src/actions/creators/common-factory.ts:24]: I'd expect a consistent ID.
Doesn't add much value to add the `-action` suffix to every single action.
```suggestion
    id: `delete-resource`,
```

File: frontend/packages/console-app/src/actions/creators/common-factory.ts
Comment (christianvogt) [inline frontend/packages/console-app/src/actions/creators/common-factory.ts:34]:
```suggestion
    id: `edit-resource`,
```

File: frontend/packages/console-app/src/actions/creators/health-checks-factory.ts
Comment (christianvogt) [inline frontend/packages/console-app/src/actions/creators/health-checks-factory.ts:41]:
```suggestion
    id: 'edit-health-checks',
```

File: frontend/packages/console-app/src/actions/creators/health-checks-factory.ts
Comment (christianvogt) [inline frontend/packages/console-app/src/actions/creators/health-checks-factory.ts:50]:
```suggestion
      verb: 'update',
```

File: frontend/packages/console-app/src/actions/providers/deployment-provider.ts
Comment (christianvogt) [inline frontend/packages/console-app/src/actions/providers/deployment-provider.ts:72]:
```suggestion
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [kindObj, relatedHPAs.length, resource, supportsHPA, t],
```

File: frontend/packages/console-shared/src/components/actions/loader/ActionsLoader.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/actions/loader/ActionsLoader.tsx:46]: You must pass the guard check first of `isResourceActionProvider` before accessing `e.properties`.

`e` is currently of type `any` when it should be of type `Extension`

**Review by christianvogt — COMMENTED** [review]
I ran into some errors creating an HPA using the new action after submitting the form. Not sure if there was something related to this change. Unfortunately the cluster went down before i could capture the error.

* Create `hello-openshift` app
* use add hpa action from topology
* immediately submit form.

File: frontend/packages/console-app/src/actions/providers/deployment-provider.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/providers/deployment-provider.ts:72]: The idea with new extension based actions was that it could easily translate the strings before evaluating each action. So each action would send a translated string rather than a labelKey.

File: frontend/packages/console-app/src/actions/providers/deployment-provider.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/providers/deployment-provider.ts:75]: Yeah, in most of the action cases we would not want to stop loading all the actions just because one of the providers got an error. It's using a generic return type for a provider because it makes it easier to type everything in extensions that has a hook to return a result.

File: frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx
Comment (rohitkrai03) [reply frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx:55]: Yeah, the action access checks are async. We do this upfront to minimise the flicker while running these access checks when menu is actually opened. So when users hover on the kebab we run these checks and memoize them. That way when action menu is actually opened it uses those memoized results rather than making another call.

File: frontend/packages/console-app/src/actions/providers/deployment-provider.ts
Comment (andrewballantyne) [reply frontend/packages/console-app/src/actions/providers/deployment-provider.ts:75]: Interesting. Once upon a time we had a concept of "optional" resource fetching. I think it was part of the Firehose if memory serves (maybe it carried over to the watch hook). Is that something we'd want to consider in this case?

I guess Extensions are sorta open-ended and you don't _really_ want to allow one to control the situation for everyone.

**Comment by rohitkrai03** [comment]
> There are duplicate actions now, right? Those in the wheelhouse of extensions and the original ones. Did we want to remove the old ones along with this change?

Yeah, we have duplicates now because not all the actions are migrated. We are using a lot of common resource actions for most of the resources. And we've not migrated actions for topology right now because it needs the migration of side panels to new extension based side panels. I would prefer to keep duplicates for now until we've moved all of our actions from older mechanism to new ones.

**Comment by rohitkrai03** [comment]
> I ran into some errors creating an HPA using the new action after submitting the form. Not sure if there was something related to this change. Unfortunately the cluster went down before i could capture the error.
> 
> * Create `hello-openshift` app
> * use add hpa action from topology
> * immediately submit form.

This is an issue on our current master as well. Seems like some issue with add HPA page. I've created a bug for this - https://issues.redhat.com/browse/ODC-6188

**Comment by andrewballantyne** [comment]
> > There are duplicate actions now, right? Those in the wheelhouse of extensions and the original ones. Did we want to remove the old ones along with this change?
> 
> @andrewballantyne Yeah, we have duplicates now because not all the actions are migrated. We are using a lot of common resource actions for most of the resources. And we've not migrated actions for topology right now because it needs the migration of side panels to new extension based side panels. I would prefer to keep duplicates for now until we've moved all of our actions from older mechanism to new ones.

Would you be okay with marking `@deprecated` on each action you migrated? It would allow us to flag things for removal without officially removing them. Plus being part of the same commit / PR would allow us to track the move (without actually moving them). All around helps with the general navigation of changes in our large repo.

```
/** @deprecated - Moving to Extensible Action, see console-app/.../actions */
export const someOldAction = () => {}
```
Or something like that.

**Comment by rohitkrai03** [comment]
> > > There are duplicate actions now, right? Those in the wheelhouse of extensions and the original ones. Did we want to remove the old ones along with this change?
> > 
> > 
> > @andrewballantyne Yeah, we have duplicates now because not all the actions are migrated. We are using a lot of common resource actions for most of the resources. And we've not migrated actions for topology right now because it needs the migration of side panels to new extension based side panels. I would prefer to keep duplicates for now until we've moved all of our actions from older mechanism to new ones.
> 
> Would you be okay with marking `@deprecated` on each action you migrated? It would allow us to flag things for removal without officially removing them. Plus being part of the same commit / PR would allow us to track the move (without actually moving them). All around helps with the general navigation of changes in our large repo.
> 
> ```
> /** @deprecated - Moving to Extensible Action, see console-app/.../actions */
> export const someOldAction = () => {}
> ```
> 
> Or something like that.

I like the idea of adding a comment like that but technically its not yet deprecated until we move all the resources that use those actions. Apart from common actions there are HPA and health check actions which are being used by `DeploymentConfig` right now, so I am not sure if its correct to label it deprecated yet. WDYT?

File: frontend/packages/console-app/src/actions/creators/hpa-factory.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/creators/hpa-factory.ts]: Its probably because it comes as the opposite of `something > 0` and it turns into `something < 1`. But I get your point. Updated.

File: frontend/packages/console-app/src/actions/providers/deployment-provider.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/providers/deployment-provider.ts]: Updated.

File: frontend/packages/console-app/src/actions/creators/common-factory.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/creators/common-factory.ts]: Updated.

File: frontend/packages/console-app/src/actions/creators/common-factory.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/creators/common-factory.ts]: Updated.

File: frontend/packages/console-app/src/actions/creators/health-checks-factory.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/creators/health-checks-factory.ts]: Updated.

File: frontend/packages/console-app/src/actions/creators/health-checks-factory.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/creators/health-checks-factory.ts]: Updated.

File: frontend/packages/console-app/src/actions/providers/deployment-provider.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/providers/deployment-provider.ts]: Updated.

File: frontend/packages/console-shared/src/components/actions/loader/ActionsLoader.tsx
Comment (rohitkrai03) [reply frontend/packages/console-shared/src/components/actions/loader/ActionsLoader.tsx]: Updated.

**Comment by andrewballantyne** [comment]
> I like the idea of adding a comment like that but technically its not yet deprecated until we move all the resources that use those actions. Apart from common actions there are HPA and health check actions which are being used by `DeploymentConfig` right now, so I am not sure if its correct to label it deprecated yet. I can probably add the comment saying migrated for `Deployment`. WDYT?

@rohitkrai03 Hmm. I definitely agree we don't want to say deprecated when it's not actually deprecated. But I think in this a close case; the old actions are a dying breed and the users of the actions are really the deprecated ones. However, the goal is that we WANT to move to the extension action model for other resources. Given there is a duplication of code, and it's the long term goal, I think we can safely call this deprecated (imo). If you want, we can clean up the "how to fix it" text. Which essentially I guess is "Update the user of the action" not so much "use this alternative action instead". 

Not sure if there is another word here we could use. I guess the only alternative is to avoid flagging as such and going with something like:
```
/**
 * Migrating to use extension actions; resource action menus should do ....xyz 
 * @see console-app/.../actions... etc
 */
export const someOldAction = () => {}
```
The `@see` can be detected by IDEs/Plugins to allow a code ref if you use a proper qualified path. Much like the `@deprecated` which usually results in a strikethrough of the identifier in the calling code.

No matter how we slice this, this model is the model we want to move to. Marking up the old actions should be done purely because of duplicated code + the intended migration of our old code, I guess we just want to figure out "the best way" to do that.

(Last moment thought, LMKWYT... Can we just hard import the new extension action Kebab item instead of using the old one? Is the code close enough/the same to just make an awkward cross package import statement for the time being instead of trying to flag/deprecate the old one)

**Comment by rohitkrai03** [comment]
@andrewballantyne I've actually added deprecated with a text saying its deprecated for deployment actions. And we cannot use the new ones to replace the old ones directly since the type is different for both.

File: frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx:80]: This should be in a useEffect. Maybe create a separate component to manage this.

File: frontend/packages/console-app/src/actions/creators/deployment-factory.ts
Comment (christianvogt) [inline frontend/packages/console-app/src/actions/creators/deployment-factory.ts:22]: Careful with these migrations when it is an object param. `configureUpdateStrategyModal` wants a prop named `deployment` and not `obj`

File: frontend/packages/console-app/src/actions/creators/deployment-factory.ts
Comment (christianvogt) [inline frontend/packages/console-app/src/actions/creators/deployment-factory.ts:35]: This action isn't updating correctly when using the action.

File: frontend/packages/console-app/src/actions/creators/common-factory.ts
Comment (christianvogt) [inline frontend/packages/console-app/src/actions/creators/common-factory.ts:73]: When testing from the deployment list page.

After editing the pod count and then reopening the modal, the number of pods is incorrect. It ends up showing the previous value still. 

Looks like there's an issue with the actions getting cached but not updated when the object changes.

File: frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx
Comment (rohitkrai03) [reply frontend/packages/console-shared/src/components/actions/LazyActionMenu.tsx]: Done.

File: frontend/packages/console-app/src/actions/creators/deployment-factory.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/creators/deployment-factory.ts]: Sorry, missed that one. Thanks for catching this.

File: frontend/packages/console-app/src/actions/creators/deployment-factory.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/creators/deployment-factory.ts:35]: Yeah, the new implementation with `ActionServiceProvider` was not updating its state for context when props changed. Updated.

File: frontend/packages/console-app/src/actions/creators/common-factory.ts
Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/creators/common-factory.ts:73]: Updated. See comment above for the main issue.

**Comment by andrewballantyne** [comment]
@rohitkrai03 is this only doing the Deployment Page? Would it be too much to do the DeploymentConfig page at the same time?

**Comment by rohitkrai03** [comment]
> @rohitkrai03 is this only doing the Deployment Page? Would it be too much to do the DeploymentConfig page at the same time?
> 
> EDIT: I ask because HPA stuff is only tied to those two workload types -- we could migrate instead of duplicating them

@andrewballantyne The idea at first was to migrate both `Deployment` and `DeploymentConfig` actions together. But the scope of the story increased due to addition of action adapters, multiple context support in action loader and lazy loaded action menu implementations. That's why I thought it would be better to just focus `Deployment` to start with. We can do a follow up to migrate deployment config and other resources.

File: frontend/packages/console-app/src/actions/creators/pvc-factory.ts
Comment (andrewballantyne) [inline frontend/packages/console-app/src/actions/creators/pvc-factory.ts:12]: I cannot find a usage for this... is this used?

Comment (rohitkrai03) [reply frontend/packages/console-app/src/actions/creators/pvc-factory.ts:12]: No, currently its not being used. I just converted all the factory actions from the old type to new type so that it would be easier to migrate other resource actions easily later.

**Review by andrewballantyne — APPROVED** [review]
Tested health checks, edit app, hpa stuff, and some of the generic actions. Everything is looking good for me. Just waiting on Christian's approval I think.

**Comment by christianvogt** [comment]
propagating from epic: 
/label px-approve
/label docs-approve

/lgtm
/label qe-approved

**Pull request merged**
